### PR TITLE
allow practical supports to be decimals

### LINKED ISSUE #2452: Change practical support amount to a decimal
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**

I guessed wrong on this one. Let's change this int to a decimal.

* **What feature or behavior is this required for?**
accurate expense tracking. oops.

* **How could we solve this issue? (Not knowing is okay!)**
* migrate the db field to a two digit decimal
* bump the input validations

* **Anything else?**
nah

### PR DESCRIPTION
**Pull request by colinxfleming**
I rule and have completed some work on Case Manager that's ready for review!

A fund noted to me that practical support amounts weren't as clean as pledge amounts, and cents mattered, so let's make this a decimal.

@elimbaum as our local expert on the practical support code, would you mind pulling this down and reviewing? 

This pull request makes the following changes:
* allows practical support amount to be a decimal in db
* ...and in form

image

It relates to the following issue #s: 
* Fixes #2452 

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -299,7 +299,7 @@
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
     t.bigint "fund_id"
-    t.integer "amount"
+    t.decimal "amount", precision: 8, scale: 2
```
Comment (elimbaum) [inline db/schema.rb:302]: when i was a wee youth i was raised to never represent currency as a float, lest you get floating point errors in your math (instead using integer cents). are we now a sufficiently advanced society that float currency is ok? is that the idea of the `precision`/`scale`?

File: app/views/practical_supports/_edit.html.erb [hunk 2]
```diff
@@ -21,7 +21,8 @@
         <%= f.number_field :amount,
                            hide_label: true,
```
Comment (elimbaum) [inline app/views/practical_supports/_edit.html.erb:22]:
```suggestion
                           value: number_to_currency(f.object.amount.to_f, unit: '', delimiter: ''),
                           hide_label: true,
```

This is a bit gross, but works. Rails's `number_to_currency` is meant to do all the work for you, but for number fields, we need to drop the unit and delimiter (`,`). This displays 2 decimal units in the edit view.

This of course doesn't work while making changes (and therefore we can't use it on the `new` partial) but I think that's a reasonable compromise. Then we can also make our tests handle currency normally. :)

Comment (elimbaum) [reply app/views/practical_supports/_edit.html.erb:22]: voila
Screen Shot 2021-12-19 at 18 35 18

File: test/system/practical_support_behaviors_test.rb [hunk 3]
```diff
@@ -21,15 +21,15 @@ class PracticalSupportBehaviorsTest < ApplicationSystemTestCase
...
+        fill_in 'Amount', with: 500.10
         check 'Confirmed'
         click_button 'Create new practical support'
       end
 
       within :css, '#practical-support-entries' do
         assert_equal 'Companion', find('#practical_support_support_type').text
         assert_equal 'Other (see notes)', find('#practical_support_source').value
-        assert_equal '500', find('#practical_support_amount').value
+        assert_equal '500.1', find('#practical_support_amount').value
```
Comment (elimbaum) [inline test/system/practical_support_behaviors_test.rb:32]: if you like the `number_to_currency` change then we can fix this!

**Review by elimbaum — APPROVED** [review]

File: app/views/practical_supports/_edit.html.erb (hunk 2 above)
Comment (elimbaum) [reply app/views/practical_supports/_edit.html.erb:22]: also i found it here: https://stackoverflow.com/questions/18898947/rails-number-field-alternative-for-decimal-values

File: db/schema.rb (hunk 1 above)
Comment (colinxfleming) [reply db/schema.rb:302]: yeah! There is mercifully a difference between floats and decimals (also called numerics) in postgres. Decimals do not do any of the unholy stuff that floats are notorious for.  https://www.postgresql.org/docs/9.5/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL

(Though honestly, there's a `money` datatype too, I'll screw around with that a bit in a commit.)

Anyway, it's not that we're an advanced enough society that float currency is okay, just that we're an advanced enough society that we no longer use float most of the time (I think).

File: app/views/practical_supports/_edit.html.erb (hunk 2 above)
Comment (colinxfleming) [reply app/views/practical_supports/_edit.html.erb:23]: minor holiday miracle in here. Thank you for doing this digging!

**Pull request merged**